Design how to configure datetime formats as part of receiver translat… (#10502)
CDCgov/prime-reportstream: prime-router/docs/features/0006-datetime-formating-as-part-of-receiver-setting.md

# Datetime formating as part of receiver translation setting

## Context

> In receiver translation setting, there is a useHighPrecisionHeaderDateTimeFormat flag that allows us to set the DateTime Percision format. When it sets true, the datetime will include milliseconds.  The DateTime Percision format can be archeived via the convertDateTimeToHL7 function in FhirPathUtils.kt evaluating the flag and return DateTime format according to the flag set to true or false.

Reviewer: ```suggestion
In the receiver translation settings, there is a useHighPrecisionHeaderDateTimeFormat flag that allows us to set the DateTime Precision format. When its set to true, the datetime will include milliseconds.  The DateTime Precision format can be achieved via the convertDateTimeToHL7 function in FhirPathUtils.kt evaluating the flag and return the DateTime format according to the flag set to true or false.
```

Author: Done

## Problem

Currently, Universal Pipeline doesn’t have ability to configure the datetime format for receivers using the the useHighPrecisionHeaderDateTimeFormat flag as COVID pipeline.

## Goal

Implement the feature to Universal Pipeline to allow us to configure datetime and work the same way as of COVID pipline receiver configutation.

## Receiver setting

```yaml
- name: lt-pdh
  description: LT Department of Health
  jurisdiction: STATE
  stateCode: LT
  receivers:
    - name: elr
      topic: covid-19
      jurisdictionalFilter: [ "matches(ordering_facility_state, LT)" ]
      translation:
        type: HL7
        useBatchHeaders: true
        receivingApplicationName: LT-PDH
        receivingApplicationOID:
        receivingFacilityName: LT-PDH
        receivingFacilityOID:
        convertDateTimesToReceiverLocalTime: false
        useHighPrecisionHeaderDateTimeFormat: true
  externalName: null
  timeZone: null
  dateTimeFormat: "OFFSET"
```

## Implementation

> •	We will pass the receiver object that contains all receiver setting to the the FHIRTranslation.kt:FhirToHl7Converter.convert method.  Inturn, it will call to the FhirPathUtils.kt:convertDateTimeToHL7.

Reviewer: ```suggestion
•	We will pass the receiver object that contains all the receiver settings to the FHIRTranslation.kt:FhirToHl7Converter.convert method.  In turn, it will call the FhirPathUtils.kt:convertDateTimeToHL7.
```

> •	We will pass the receiver object that contains all receiver setting to the the FHIRTranslation.kt:FhirToHl7Converter.convert method.  Inturn, it will call to the FhirPathUtils.kt:convertDateTimeToHL7.

Reviewer: How will we know when to call the convertDateTimeToHL7 in FhirToHl7Converter.convert?

Author: There is a long thread calling from FhirToHl7Converter.convert to FhirPathUtils.kt:convertDateTimeToHL7 during FHIRTranslator.kt:getHL7MessageFromBundle is calling.

> •	To formate dateTime, convertDateTimeToHL7 will use the DateUtilities.kt:formatDateForReceiver function as COVID pipeline does.   All attributes needed to format datetime are in the receiver object.

Reviewer: We're planning to migrate the conversion library to its own repo and we won't be able to reference code that is outside of it. We would have to move the DateUtilities.kt:formatDateForReceiver logic inside the conversion library. We already have a function to convert timezones in the FHIR path utils, maybe that can be reused.

Author: Yeah, if we already have a function to convert timezone in the FHIR path utils, we should reuse that one.  I love to reuse code.

> •	To formate dateTime, convertDateTimeToHL7 will use the DateUtilities.kt:formatDateForReceiver function as COVID pipeline does.   All attributes needed to format datetime are in the receiver object.

Reviewer: ```suggestion
•	To format dateTime, convertDateTimeToHL7 will use the DateUtilities.kt:formatDateForReceiver function as the COVID pipeline does.   All attributes needed to format datetime are in the receiver object.
```
